metagraph/src/graph/representation/succinct/boss.cpp

```diff
@@ -2273,6 +2288,172 @@ void call_paths(const BOSS &boss,
     }
 }
 
+/**
+ * Updates #terminal and #near_terminal based on the given path.
+ * One terminal node is set every max_length nodes in the path, and all nodes before it
+ * are marked as near_terminal.
+ * The last node in the path is marked as terminal if:
+ *  1. The path length is an exact multiple of max_length, OR
+ *  2. The last node in the path is a dead-end, OR
+ *  3. The last node into a path merges into a node that is neither terminal nor near
+ *     terminal
+ *  If the last node in the path is not terminal then #callback is invoked with the anchor
+ *  node against which the last node should be diffed.
+ */
+void update_terminal_bits(
+        const BOSS::Call<std::vector<edge_index> &&, std::optional<edge_index>> &callback,
+        uint64_t max_length,
+        edge_index *merge_edge,
+        std::vector<edge_index> &&path,
+        sdsl::bit_vector *terminal,
+        sdsl::bit_vector *near_terminal) {
+    uint64_t i = 0;
+    constexpr bool async = true;
+    for (i = 0; i + max_length <= path.size(); i += max_length) {
+        for (uint64_t j = i; j < i + max_length - 1UL; ++j) {
+            set_bit(near_terminal->data(), path[j], async);
+        }
+        set_bit(terminal->data(), path[i + max_length - 1], async);
+    }
+
+    if ( path.size() % max_length == 0 ) { // last node is terminal
+        callback(std::move(path), nullopt);
+        return;
+    }
+
+    // mark the last node in the path as terminal if
+    // 1. there are no outgoing edges, OR
+    // 2. we merge into a node that is neither terminal nor near terminal
+    bool merge_terminal = merge_edge && fetch_bit(terminal->data(), *merge_edge);
+    bool merge_near_terminal = merge_terminal
```

Comment:
Hmmm, I think it must be a merge. If it's a simple fork, then we could just pick the last outgoing edge and continue the traversal. But since we had to stop the traversal, it means that the last outgoing edge was already visited, and that's only possible if there is a merge at this node.

Reply:
The last outgoing may happen to be already visited (in fact all outgoing may be visited to this moment). Remember the second block starts traversals at forks?

Reply:
Ah, yes, I missed the fact that we start traversals at forks. Renamed.